### PR TITLE
Data_Enrichment Module

### PR DESCRIPTION
**Pull request by pamodaDilranga**
**PR Request: Enhance Data Enrichment with Paraphrasing and Translation Capabilities**
Overview:
This pull request introduces enhancements to the data enrichment capabilities of our project, focusing on integrating paraphrasing and translation functionalities using transformer models.

Changes:
Data Enrichment (data_enrichment.py):

Introduced the DataEnrichment class to handle input text enrichment.
Integrated Paraphraser and Translator classes to provide enhanced functionality based on language detection and transformation.
Paraphraser (paraphraser.py):

Implemented the Paraphraser class using Hugging Face's AutoModelForSeq2SeqLM and AutoTokenizer for paraphrasing tasks.
Generates multiple paraphrases for input text leveraging transformer-based seq2seq models.
Translator (translator.py):

Implemented the Translator class using Hugging Face's MarianMTModel and MarianTokenizer for translation tasks.
Supports translation between English and multiple languages including Estonian, Russian, Polish, and Finnish.

**Functionality**:
Data Enrichment:

Detects the language of input text and determines if it's supported (English, Estonian, Russian, Polish, Finnish).
If supported:
Directly generates paraphrases using Paraphraser for English text.
Translates non-English text to English using Translator, generates paraphrases, and translates them back to the original language.
If unsupported:
Logs an error message indicating the language is not supported.
Paraphraser:

Utilizes transformer-based seq2seq models to generate paraphrases of input text.
Configurable parameters such as number of beams, diversity penalty, and maximum length for paraphrase generation.
Translator:

Implements translation between English and multiple languages using transformer-based translation models.
Handles tokenization, model loading, and translation generation using pretrained models from Hugging Face's model hub.
Example Usage:
python
Copy code
```
from data_enrichment import DataEnrichment

enricher = DataEnrichment()

# Example text and language
text = "Hello, I hope this email finds you well. I am writing to inquire about the status of my application."
lang = 'en'  # English

# Enrich data
enriched_data = enricher.enrich_data(text, lang)

if enriched_data:
    print("Enriched Data:")
    for i, data in enumerate(enriched_data, 1):
        print(f"{i}. {data}")
```
Impact:
Improved Data Utilization: Enhances the utility of our project by providing enriched data through paraphrasing and translation, improving natural language understanding and interoperability.
Flexibility: Supports a variety of languages, enabling broader application scenarios and user engagement.
Tests:
Included unit tests to validate paraphrasing and translation functionalities across supported languages.
Tested edge cases such as unsupported languages to ensure proper error handling and user feedback.
Next Steps:
Enhance error handling for edge cases and unsupported languages.
Explore additional transformer models or fine-tuning strategies for improved paraphrasing and translation quality.

### CODE REVIEW

File: src/data_enrichment/data_enrichment.py [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+
+        return paraphrases
+
+if __name__ == "__main__":
+    enricher = DataEnrichment()
+
+    sentences = [
+        ("Hello, I hope this email finds you well. I am writing to inquire about the status of my application. I submitted all required documents last month but have not received any update since then. Could you please provide me with the current status?", 'en'),
+        ("Tere, ma loodan, et see e-kiri leiab teid hästi. Ma kirjutan päringu staatuse minu taotluse. Ma esitasin kõik nõutavad dokumendid möödunud kuu, kuid ma pole saanud ühtegi värskendust sellest ajast. Kas saaksite palun mulle hetkeseisu kohta teavet anda?", 'et'),  # Estonian
+        ("Привет, надеюсь, что это письмо находит вас в порядке. Я пишу с запросом о статусе моего заявления. Я подал все необходимые документы в прошлом месяце, но с тех пор не получил никаких обновлений. Не могли бы вы, пожалуйста, предоставить мне текущий статус?", 'ru'),  # Russian
```
Comment (Thirunayan22) [inline src/data_enrichment/data_enrichment.py:35]: It's better if we create a seperate JSON file and the examples to that file and read the file @pamodaDilranga . Also we need to make it a practice to never include examples like this directly in the main function of class file. Looks unclean.

File: src/data_enrichment/data_enrichment.py [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+        ("Cześć, mam nadzieję, że to e-mail znajduje cię w dobrym zdrowiu. Piszę z zapytaniem o status mojej aplikacji. Złożyłem wszystkie wymagane dokumenty w zeszłym miesiącu, ale od tego czasu nie otrzymałem żadnej aktualizacji. Czy moglibyście podać mi bieżący status?", 'pl'),  # Polish
+        ("Hei, toivottavasti tämä sähköposti tavoittaa teidät hyvin. Kirjoitan tiedustelun tilastani hakemukseni suhteen. Lähetin kaikki tarvittavat asiakirjat viime kuussa, mutta en ole sen jälkeen saanut päivityksiä. Voisitteko antaa minulle ajantasaisen tilanteen?", 'fi')  # Finnish
+    ]
+
+
+    for sentence, lang in sentences:
+        print(f"Original sentence ({lang}): {sentence}")
+        enriched_data = enricher.enrich_data(sentence, lang)
+        if enriched_data:
```
Comment (Thirunayan22) [inline src/data_enrichment/data_enrichment.py:44]: Instead of testing the functionality here, we need create a separate folder called tests and under that create a file called test_data_enrichment and include this test over there. Make sure that functionality works even after moving @pamodaDilranga

File: src/data_enrichment/data_enrichment.py [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+    def __init__(self):
+        self.translator = Translator()
+        self.paraphraser = Paraphraser()
+
+    def enrich_data(self, text, lang):
+        supported_languages = ['en', 'et', 'ru', 'pl', 'fi']
+
+        if lang not in supported_languages:
+            print(f"Unsupported language: {lang}")
+            return None
```
Comment (Thirunayan22) [inline src/data_enrichment/data_enrichment.py:14]: Never return None, return a error message defined in a JSON file. This can be a generic error message and remove print statement and replace it with the logging library,

File: src/data_enrichment/data_enrichment.py [hunk 4]
```diff
@@ -0,0 +1,48 @@
+from translator import Translator
+from paraphraser import Paraphraser
+
+class DataEnrichment:
+    def __init__(self):
+        self.translator = Translator()
+        self.paraphraser = Paraphraser()
+
+    def enrich_data(self, text, lang):
```
Comment (Thirunayan22) [inline src/data_enrichment/data_enrichment.py:9]: add parameter types and return types

Comment (Thirunayan22) [reply src/data_enrichment/data_enrichment.py:9]: In every function

File: src/data_enrichment/paraphraser.py [hunk 5]
```diff
@@ -0,0 +1,35 @@
+from transformers import AutoTokenizer, AutoModelForSeq2SeqLM
+
+class Paraphraser:
+    def __init__(self):
+        self.model_name = "humarin/chatgpt_paraphraser_on_T5_base"
```
Comment (Thirunayan22) [inline src/data_enrichment/paraphraser.py:5]: make this a parameter in config file, we shouldn't ideally hard code this @pamodaDilranga

File: src/data_enrichment/translator.py [hunk 6]
```diff
@@ -0,0 +1,47 @@
+from transformers import MarianMTModel, MarianTokenizer
+
+class Translator:
+    def __init__(self):
+        self.models = {
+            'et-en': ('Helsinki-NLP/opus-mt-et-en', 'Helsinki-NLP/opus-mt-en-et'),
+            'ru-en': ('Helsinki-NLP/opus-mt-ru-en', 'Helsinki-NLP/opus-mt-en-ru'),
+            'pl-en': ('Helsinki-NLP/opus-mt-pl-en', 'Helsinki-NLP/opus-mt-en-pl'),
```
Comment (Thirunayan22) [inline src/data_enrichment/translator.py:8]: This entire dictionary should be in a json file , read it from there @pamodaDilranga

File: src/data_enrichment/translator.py [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+            'fr-pl': ('Helsinki-NLP/opus-mt-fr-pl', 'Helsinki-NLP/opus-mt-pl-fr'),
+        }
+        self.tokenizers = {}
+        self.models_instances = {}
+
+        for key, (model_name, reverse_model_name) in self.models.items():
+            self.tokenizers[key] = MarianTokenizer.from_pretrained(model_name)
+            self.models_instances[key] = MarianMTModel.from_pretrained(model_name)
+            reverse_key = f"{key.split('-')[1]}-{key.split('-')[0]}"
+            if reverse_model_name != 'Helsinki-NLP/opus-mt-en-pl':
```
Comment (Thirunayan22) [inline src/data_enrichment/translator.py:20]: Same here, details like this should be picked from a config file. Like "models.config" @pamodaDilranga

Comment (Thirunayan22) [reply src/data_enrichment/translator.py:20]: or models.json

**Review by Thirunayan22 — CHANGES_REQUESTED** [review]

File: src/data_enrichment/data_enrichment.py (hunk 1 above)
Comment (pamodaDilranga) [reply src/data_enrichment/data_enrichment.py]: Created unit tests

File: src/data_enrichment/data_enrichment.py (hunk 2 above)
Comment (pamodaDilranga) [reply src/data_enrichment/data_enrichment.py]: Created unit tests

File: src/data_enrichment/data_enrichment.py (hunk 3 above)
Comment (pamodaDilranga) [reply src/data_enrichment/data_enrichment.py]: Changed it to return empty list

File: src/data_enrichment/data_enrichment.py (hunk 4 above)
Comment (pamodaDilranga) [reply src/data_enrichment/data_enrichment.py]: updated all the functions with parameter types and return info

File: src/data_enrichment/paraphraser.py (hunk 5 above)
Comment (pamodaDilranga) [reply src/data_enrichment/paraphraser.py]: Created config files and updated the code

File: src/data_enrichment/translator.py (hunk 6 above)
Comment (pamodaDilranga) [reply src/data_enrichment/translator.py]: Created config files and updated the code

File: src/data_enrichment/translator.py (hunk 7 above)
Comment (pamodaDilranga) [reply src/data_enrichment/translator.py:18]: Created config files and updated the code

**Review by Thirunayan22 — APPROVED** [review]
Looks Good 👍
